deptagency/algomart
# Setup nx

Replaces custom build configurations with [Nx](https://nx.dev/).

## Notes

- Most changes come from the `package-lock.json` file. 
- I recommend installing the `nx` CLI: `npm install --global nx`
- To run all projects (in dev mode): `npm start`
- To build all projects: `npm run build`
- To test all projects: `npm test`
- To lint all projects: `npm run lint`
- For individual commands, use commands like `nx serve api` or `nx test web`
- CMS database migrations have been removed in favor of the built-in snapshot tool. You can try it out with a **blank** database: `nx import cms`
- To export changes to the CMS data model, run `nx export cms` and commit the updated `snapshot.yml` file.
- CMS seed script still works the same as before, for now.
- Basic dockerfiles for the API and Web projects are provided, but will need to extra checks for security hardening.
- ESLint rules _should_ work the same
- No major source code changes. Some TypeScript related errors were fixed.
- Building all projects result in build artifacts in the `dist` directory. The relevant projects include a generated `package.json` file with its runtime dependencies. _Some_ dependencies might be missing if they're not directly imported in the related source code. See example of that in the `docker/api/Dockerfile` image.
- The `web-e2e` project is a basic Cypress setup to be used for end-to-end testing the Web project.
- Use separate databases for the CMS and API projects due to how the snapshot feature works.
